liblocation/src/main/java/com/mapbox/android/core/location/GpxParser.java

```diff
@@ -0,0 +1,103 @@
+package com.mapbox.android.core.location;
+
+import android.location.Location;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NamedNodeMap;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Locale;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+
+class GpxParser {
+
+  private static final String TAG_TRACK_POINT = "trkpt";
+  private static final String TAG_TIME = "time";
+  private static final String ATTR_LATITUDE = "lat";
+  private static final String ATTR_LONGITUDE = "lon";
+  private static final String GPX_LOCATION_NAME = "GPX Generated Location";
+  private static final String DATE_FORMAT_PATTERN = "yyyy-MM-dd'T'HH:mm:ss'Z'";
+
+  private SimpleDateFormat dateFormat = new SimpleDateFormat(DATE_FORMAT_PATTERN, Locale.getDefault());
+
+  @Nullable
+  List<Location> parseGpx(InputStream inputStream) throws ParserConfigurationException,
+    SAXException, IOException, ParseException {
+
+    DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+    DocumentBuilder documentBuilder = documentBuilderFactory.newDocumentBuilder();
+    Document document = documentBuilder.parse(inputStream);
+    Element elementRoot = document.getDocumentElement();
+
+    NodeList trackPointNodes = elementRoot.getElementsByTagName(TAG_TRACK_POINT);
+    if (trackPointNodes == null || trackPointNodes.getLength() == 0) {
+      return null; // Gpx trace did not contain correct tagging
+    }
+
+    try {
+      return createGpxLocationList(trackPointNodes);
```

Comment:
can we use try with resources instead? can the IOException be handled by the end-developer themselves?